driver-core/src/main/com/mongodb/internal/connection/OidcAuthenticator.java

```diff
@@ -0,0 +1,656 @@
+    }
+
+    @Override
+    public void reauthenticate(final InternalConnection connection) {
+        fallbackState = FallbackState.INITIAL;
+        authLock(connection, connection.getDescription());
+    }
+
+    @Override
+    public void authenticate(final InternalConnection connection, final ConnectionDescription connectionDescription) {
+        // method must only be called during original handshake; fail otherwise
+        assertFalse(connection.opened());
+        // this method "wraps" the default authentication method in custom OIDC retry logic
+        String accessToken = getValidCachedAccessToken();
+        if (accessToken != null) {
+            try {
+                authenticateUsing(connection, connectionDescription, (bytes) -> prepareTokenAsJwt(accessToken));
+            } catch (MongoSecurityException e) {
+                if (triggersRetry(e)) { // TODO-OIDC-x unclear how to provide test coverage for this
+                    authLock(connection, connectionDescription);
+                } else {
+                    throw e;
+                }
+            }
+        } else {
+            authLock(connection, connectionDescription);
+        }
+    }
+
+    private static boolean triggersRetry(@Nullable final Throwable t) {
+        if (t instanceof MongoSecurityException) {
+            MongoSecurityException e = (MongoSecurityException) t;
+            Throwable cause = e.getCause();
+            if (cause instanceof MongoCommandException) {
+                MongoCommandException commandCause = (MongoCommandException) cause;
+                return commandCause.getErrorCode() == 18;
+            }
+        }
+        return false;
+    }
+
+
+    private void authenticateUsing(
+            final InternalConnection connection,
+            final ConnectionDescription connectionDescription,
+            final Function<byte[], byte[]> evaluateChallengeFunction) {
+        this.evaluateChallengeFunction = evaluateChallengeFunction;
+        super.authenticate(connection, connectionDescription);
+    }
+
+    private void authLock(final InternalConnection connection, final ConnectionDescription connectionDescription) {
+        MongoCredentialWithCache mongoCredentialWithCache = getMongoCredentialWithCache();
+        mongoCredentialWithCache.withOidcLock(() -> {
+            while (true) {
+                try {
+                    authenticateUsing(connection, connectionDescription, (challenge) -> evaluate(challenge));
+                    break;
+                } catch (MongoSecurityException e) {
+                    OidcCacheEntry cacheEntry = mongoCredentialWithCache.getOidcCacheEntry();
+                    if (triggersRetry(e)) {
+                        prepareRetry(e, cacheEntry);
+                    } else {
+                        throw e;
+                    }
+                }
+            }
+            return null;
+        });
+    }
```

Comment:
`prepareRetry(e, cacheEntry)` sometimes throws `e`, but the construct

```java
if (triggersRetry(e)) {
    prepareRetry(e, cacheEntry);
} else {
    throw e;
}
```

suggests (at least to me, when I read) that it does not, and it only uses `e` to prepare the next attempt. Here is the suggested change as well as the example of how `authLockAsync` can be implemented using `RetryingAsyncCallbackSupplier` (if we do this, we should replace the `while` loop in the sync code with `RetryingSyncSupplier`): https://github.com/stIncMale/mongo-java-driver/commit/2db23a89948e68f4dd7515d2c777aab3b2d501d3.

P.S. `AsyncCallbackLoop` can't be used because it emulates `while(true)`, but does not provide any mechanism for emulating `catch` blocks.
P.P.S. `fallbackState` can technically continue to live on `OidcAuthenticator` as opposed to be attached to `RetryState` because `OidcAuthenticator` is not used concurrently. Attaching `fallbackState` to `RetryState` still better expresses the intent in my opinion.